# Trash Compactor

AOC Day 6 Part 1

```
123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
```

So, this worksheet contains four problems:

- $123 \times 45 \times 6 = 33210$
- $328 + 64 + 98 = 490$
- $51 \times 387 \times 215 = 4243455$
- $64 + 23 + 314 = 401$

To check their work, cephalopod students are given the grand total of adding together all of the answers to the individual problems. In this worksheet, the grand total is $33210 + 490 + 4243455 + 401 = 4277556$

This is all about processing the input data. Ideally, a vector needs to be initialised for each operation.

In [232]:
input = [
    "123 328  51 64 ",
    " 45 64  387 23 ",
    "  6 98  215 314",
    "*   +   *   +  "
]

matrix = [line.split() for line in input]
print(matrix)

[['123', '328', '51', '64'], ['45', '64', '387', '23'], ['6', '98', '215', '314'], ['*', '+', '*', '+']]


In [233]:
for col in range(len(matrix[0])):
    for row in range(len(matrix)):
        print(matrix[row][col])

123
45
6
*
328
64
98
+
51
387
215
*
64
23
314
+


That is in fact what we are looking for.

In [234]:
help(list.pop)

Help on method_descriptor:

pop(self, index=-1, /) unbound builtins.list method
    Remove and return item at index (default last).

    Raises IndexError if list is empty or index is out of range.



In [235]:
from functools import reduce

result = 0
for col in range(len(matrix[0])):
    vector = []
    for row in range(len(matrix)):
        vector.append(matrix[row][col])
    operator = "*" if vector.pop() == "*" else "+"
    if operator == "*":
        result += reduce(lambda x, y: int(x)*int(y), vector)
    elif operator == "+":
        result += reduce(lambda x, y: int(x) + int(y), vector)
    else:
        raise ValueError()
    
print(result)
    

4277556


This is in fact the correct answer.

Let's do it for our input.

In [236]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    matrix = [line.split() for line in input]

    print(matrix)

[['886', '63', '27', '258', '98', '318', '99', '975', '7', '6393', '947', '87', '23', '765', '35', '1', '6415', '4', '5', '882', '7', '475', '74', '598', '4', '47', '66', '9233', '6', '669', '2', '3', '6', '92', '77', '4', '6', '64', '89', '989', '436', '833', '946', '5927', '398', '2', '262', '594', '752', '79', '3', '3', '44', '117', '149', '74', '119', '48', '17', '762', '621', '81', '3576', '23', '259', '141', '51', '312', '16', '255', '6', '925', '7', '1', '395', '61', '61', '34', '277', '1366', '44', '8', '32', '26', '932', '61', '48', '29', '693', '3', '27', '89', '9', '618', '95', '693', '3', '3', '11', '7', '366', '2455', '3', '11', '4', '59', '9', '9975', '81', '2', '71', '989', '697', '8636', '12', '16', '889', '78', '53', '23', '321', '33', '29', '129', '81', '18', '161', '78', '7365', '8', '9515', '127', '94', '383', '6', '17', '2', '389', '724', '37', '5227', '64', '55', '5', '1', '9', '983', '36', '194', '377', '476', '938', '81', '39', '76', '3', '9', '935', '16', '16',

In [237]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    matrix = [line.split() for line in input]

    result = 0
    for col in range(len(matrix[0])):
        vector = []
        for row in range(len(matrix)):
            vector.append(matrix[row][col])
        operator = "*" if vector.pop() == "*" else "+"
        if operator == "*":
            result += reduce(lambda x, y: int(x)*int(y), vector)
        elif operator == "+":
            result += reduce(lambda x, y: int(x) + int(y), vector)
        else:
            raise ValueError()
        
print(result)

5873191732773


$5873191732773$ _is_ the correct answer!

---

Part II is a bit difficult.



Here's the example worksheet again:

```
123 328  51 64 
 45 64  387 23 
  6 98  215 314
*   +   *   +  
```

Reading the problems right-to-left one column at a time, the problems are now quite different:

- The rightmost problem is $4 + 431 + 623 = 1058$
- The second problem from the right is $175 \times 581 \times 32 = 3253600$
- The third problem from the right is $8 + 248 + 369 = 625$
- Finally, the leftmost problem is $356 \times 24 \times 1 = 8544$

Sum is $1058 + 3253600 + 625 + 8544 = 3263827$.

It is clear as day that the `split()` logic **would not** work anymore. Or at least, it needs to be changed.

Since the operators are always aligned at the most-significant bit, we can imagine the following:

If $i$ was the index of the operator $o_1$, and $j$ was the next operator in the sequence $o_2$, we can decide that the split occurs at $(j - i) - 1$.This is the case because $(j - i)$ would give the distance between the operators, and $-1$ is required because there is a whitespace seperating two columns.

In [238]:
input = [
    "123 328  51 64 ",
    " 45 64  387 23 ",
    "  6 98  215 314",
    "*   +   *   +  "
]

operator_sequence = input[-1]
operator_sequence

'*   +   *   +  '

In [239]:
# Check the indexes in the sequence where an operator exists.

operator_indexes = []

for j in range(len(operator_sequence)):
    if operator_sequence[j] == "*" or operator_sequence[j] == "+":
        operator_indexes.append(j if j == 0 else j - 1)

operator_indexes

[0, 3, 7, 11]

Let's take the first row.

In [240]:
input[0]

'123 328  51 64 '

In [241]:
[input[0][i:j] for i,j in zip(operator_indexes, operator_indexes[1:]+[None])]

['123', ' 328', '  51', ' 64 ']

In [242]:
A = [t for t in zip(operator_indexes, operator_indexes[1:]+[None])]

In [243]:
for i in range(1, len(A)):
    x, y = A[i]
    A[i] = (x + 1, y)

A

[(0, 3), (4, 7), (8, 11), (12, None)]

In [244]:
[input[0][i:j] for i,j in A]

['123', '328', ' 51', '64 ']

Let us wrap the splitting behaviour into a `split` function.

In [245]:
def split(row: str, operator_row: str) -> list[str]:
    # Prepare indexes
    indexes = []
    for j in range(len(operator_row)):
        if operator_row[j] == "*" or operator_row[j] == "+":
            indexes.append(j if j == 0 else j - 1)
    # Split with the indexes
    intermediary_indexes = [t for t in zip(indexes, indexes[1:]+[None])]
    for i in range(1, len(intermediary_indexes)):
        x, y = intermediary_indexes[i]
        intermediary_indexes[i] = (x + 1, y)
    return [row[i:j] for i,j in intermediary_indexes]

Testing it with the first row.

In [246]:
split(input[0], input[-1])

['123', '328', ' 51', '64 ']

Let's prepare the matrix now.

In [247]:
def f(input):
    ops = input.pop()
    op_vector = ops.split()
    matrix = [split(input[row], operator_row=ops) for row in range(0, len(input))]
    result = 0

    for col in range(0, len(matrix[0])):
        vector = []
        operator = op_vector[col]
        for row in range(0, len(matrix)):
            vector.append(matrix[row][col])

        # i is a pointer iterating across decimal places.
        col_ans_vector = []
        for i in reversed(range(len(vector[0]))):
            number = ""
            # num is a pointer iterating across the numbers in the column vector.
            for num in range(len(vector)):
                number += vector[num][i]

            print(f"Candidate number: {number}")
            
            col_ans_vector.append(int(number))
        
        print(f"Operator = {operator}")
            
        if operator == "*":
            result += reduce(lambda x, y: int(x)*int(y), col_ans_vector)
        elif operator == "+":
            result += reduce(lambda x, y: int(x) + int(y), col_ans_vector)
        else:
            raise ValueError()

    return result

input = [
    "123 328  51 64 ",
    " 45 64  387 235",
    "  6 98  215 314",
    "*   +   *   +  "
]
print(f"Input = {input}")
f(input)

Input = ['123 328  51 64 ', ' 45 64  387 235', '  6 98  215 314', '*   +   *   +  ']
Candidate number: 356
Candidate number: 24 
Candidate number: 1  
Operator = *
Candidate number: 8  
Candidate number: 248
Candidate number: 369
Operator = +
Candidate number: 175
Candidate number: 581
Candidate number:  32
Operator = *
Candidate number:  54
Candidate number: 431
Candidate number: 623
Operator = +


3263877

$3263827$ is infact, the correct answer!

Let's invoke the function `f` for our `input.txt`.

In [248]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]
    print(f(input))

Candidate number: 62  
Candidate number: 855 
Candidate number: 8668
Operator = *
Candidate number: 3936
Candidate number: 6448
Candidate number:   16
Operator = +
Candidate number: 7465
Candidate number: 2139
Candidate number:    6
Operator = +
Candidate number: 8789
Candidate number: 59  
Candidate number: 2   
Operator = *
Candidate number: 822 
Candidate number: 9744
Operator = +
Candidate number: 8599
Candidate number: 1328
Candidate number: 3346
Operator = +
Candidate number: 985 
Candidate number: 9252
Operator = *
Candidate number: 5973
Candidate number: 7627
Candidate number: 9233
Operator = +
Candidate number: 7346
Candidate number:  249
Operator = *
Candidate number: 3851
Candidate number: 9473
Candidate number: 3671
Candidate number: 61  
Operator = +
Candidate number: 788 
Candidate number: 495 
Candidate number: 9657
Operator = +
Candidate number: 7661
Candidate number: 8474
Operator = *
Candidate number: 3856
Candidate number: 2689
Operator = *
Candidate number: 541 
Can

The last canidate number looks like it's missing the last `513` value.

In [249]:
input = [
    "58 ",
    "665",
    "911",
    "343",
    "*  "
]
f(input)

Candidate number:  513
Candidate number: 8614
Candidate number: 5693
Operator = *


25157264526

It looks like an issue with reading the file.

In [250]:
with open(file="input.txt") as file:
    input = [line.rstrip() for line in file]

    print(input[0][-1])

8


So, the issue finally turned out to be the `rstrip()` function in the `read()` function. It removes trailing whitespace from the first column which screws up one candidate number.

In [253]:
with open(file="input.txt") as file:
    input = [line.rstrip("\n") for line in file]
    print(f(input))

Candidate number: 62  
Candidate number: 855 
Candidate number: 8668
Operator = *
Candidate number: 3936
Candidate number: 6448
Candidate number:   16
Operator = +
Candidate number: 7465
Candidate number: 2139
Candidate number:    6
Operator = +
Candidate number: 8789
Candidate number: 59  
Candidate number: 2   
Operator = *
Candidate number: 822 
Candidate number: 9744
Operator = +
Candidate number: 8599
Candidate number: 1328
Candidate number: 3346
Operator = +
Candidate number: 985 
Candidate number: 9252
Operator = *
Candidate number: 5973
Candidate number: 7627
Candidate number: 9233
Operator = +
Candidate number: 7346
Candidate number:  249
Operator = *
Candidate number: 3851
Candidate number: 9473
Candidate number: 3671
Candidate number: 61  
Operator = +
Candidate number: 788 
Candidate number: 495 
Candidate number: 9657
Operator = +
Candidate number: 7661
Candidate number: 8474
Operator = *
Candidate number: 3856
Candidate number: 2689
Operator = *
Candidate number: 541 
Can

$11386445308378$ is the correct answer!